cpp/tests/streams/io/csv_test.cpp

```diff
@@ -0,0 +1,114 @@
+#include <sstream>
+#include <string>
+#include <vector>
+
+auto const temp_env = static_cast<cudf::test::TempDirTestEnvironment*>(
+  ::testing::AddGlobalTestEnvironment(new cudf::test::TempDirTestEnvironment));
+
+class CSVTest : public cudf::test::BaseFixture {};
+
+template <typename T>
+inline auto random_values(size_t size)
+{
+  std::vector<T> values(size);
+
+  using T1 = T;
+  using uniform_distribution =
+    typename std::conditional_t<std::is_same_v<T1, bool>,
+                                std::bernoulli_distribution,
+                                std::conditional_t<std::is_floating_point_v<T1>,
+                                                   std::uniform_real_distribution<T1>,
+                                                   std::uniform_int_distribution<T1>>>;
+
+  static constexpr auto seed = 0xf00d;
+  static std::mt19937 engine{seed};
+  static uniform_distribution dist{};
+  std::generate_n(values.begin(), size, [&]() { return T{dist(engine)}; });
+
+  return values;
+}
+
+TEST_F(CSVTest, CSVReader)
+{
+  constexpr auto num_rows = 10;
+  auto int8_values        = random_values<int8_t>(num_rows);
+  auto int16_values       = random_values<int16_t>(num_rows);
+  auto int32_values       = random_values<int32_t>(num_rows);
+  auto int64_values       = random_values<int64_t>(num_rows);
+  auto uint8_values       = random_values<uint8_t>(num_rows);
+  auto uint16_values      = random_values<uint16_t>(num_rows);
+  auto uint32_values      = random_values<uint32_t>(num_rows);
+  auto uint64_values      = random_values<uint64_t>(num_rows);
+  auto float32_values     = random_values<float>(num_rows);
+  auto float64_values     = random_values<double>(num_rows);
+
+  auto filepath = temp_env->get_temp_dir() + "MultiColumn.csv";
+  {
+    std::ostringstream line;
+    for (int i = 0; i < num_rows; ++i) {
+      line << std::to_string(int8_values[i]) << "," << int16_values[i] << "," << int32_values[i]
+           << "," << int64_values[i] << "," << std::to_string(uint8_values[i]) << ","
+           << uint16_values[i] << "," << uint32_values[i] << "," << uint64_values[i] << ","
+           << float32_values[i] << "," << float64_values[i] << "\n";
+    }
+    std::ofstream outfile(filepath, std::ofstream::out);
+    outfile << line.str();
+  }
+
+  cudf::io::csv_reader_options in_opts =
+    cudf::io::csv_reader_options::builder(cudf::io::source_info{filepath})
+      .header(-1)
+      .dtypes({cudf::data_type{cudf::type_id::INT8},
+               cudf::data_type{cudf::type_id::INT16},
+               cudf::data_type{cudf::type_id::INT32},
+               cudf::data_type{cudf::type_id::INT64},
+               cudf::data_type{cudf::type_id::UINT8},
+               cudf::data_type{cudf::type_id::UINT16},
+               cudf::data_type{cudf::type_id::UINT32},
+               cudf::data_type{cudf::type_id::UINT64},
+               cudf::data_type{cudf::type_id::FLOAT32},
+               cudf::data_type{cudf::type_id::FLOAT64}});
```

Comment:
If we don't pass the data types the reader will launch an extra kernel to infer them, so we should not include this option :)